### ISSUE TITLE
CompileString/RunString/Script errors upon a NUL char

### ISSUE THREAD

**Issue by Cheatoid** [Facepunch/garrysmod-issues#4822]
### Details
As the title says.

### Steps to reproduce
0. [Download the example script](https://github.com/Facepunch/garrysmod-issues/files/5932496/null.txt).
1. Run the script either via `CompileString`/`CompileFile`/`RunString`/`lua_openscript`...
2. Observe the script error:
    `unfinished long string near '<eof>'`

My hypothesis that this is because of all C API (luashared/engine/luajit) are using `const char*` all over the place, gmod ain't compiled with `UNICODE` switch therefore the "`code`" ends up terminated upon a NUL char.

This is pretty annoying issue for me because I want to embed the binary data within code ***without*** having to escape NULs as `\000`...

**Comment by Kefta** [comment]
This is actually to prevent bytecode from being loaded. It's a bad hack that needs to be solved properly.

**Comment by robotboy655** [comment]
> This is actually to prevent bytecode from being loaded

Nah, it's just not handled properly. At least with CompileString proper bytecode will still not load, I have tested it.

**Comment by Cheatoid** [comment]
...I thought bytecode-loading was completely stripped rather than hacked-away... 😆

In my testing, `RunString`/`CompileString`/`lua_openscript`/whatever wrappers, it all boils down to this ultimate function:
`lua_load` (or rather `luaL_loadbufferx`, which is sometimes inlined...)

The clue is on the docs: <https://www.lua.org/manual/5.4/manual.html#lua_load>
> `lua_load` **automatically detects** whether the chunk is **text or binary** and loads it accordingly.
> The string mode works as in function `load`, with the addition that a **`NULL` value is equivalent to the string "bt"**.

The value of the `mode` always seems to be `NULL` (0) in gmod, which means it will try to automatically detect the mode.
So...
https://github.com/LuaJIT/LuaJIT/blob/ec6edc5c39c25e4eb3fca51b753f9995e97215da/src/lj_load.c#L37
Get rid of the automatic-mode-detection part?
Fully strip the bytecode-loading capability?
And possibly, always provide `"t"` string as the `mode` parameter value? (notice no `b`)
(not sure if any of this is the correct approach, just throwing some ideas; perhaps do all of them? 😄)

Then it is just a matter of providing the correct value to the `sz` parameter to fix NUL issue, I believe.

**Comment by ThatLing** [comment]
As a temporary solution, you can use '\0' which is 2 bytes instead of 1

**Comment by thegrb93** [comment]
That's the actual solution rather

**Comment by Cheatoid** [comment]
Smarty... Think again.
It is not 2 bytes. I have to use 4 bytes. Because what if there are digits ahead of the `\0`
Guess what, it will treat them as part of the escape. Example:
`..... \04rekt .....`
The only time you can just use `\0` is when you know that there are no digits ahead of it.
Hence why I'd have to do `\000` to ensure proper NUL escaping.

**Comment by Cheatoid** [comment]
Moreover you can't use escaping in long string `[[ long string ]]` at all.
Which means I'd also have to escape those:
`\" \\ \n \r` ... which is terrible, but it is a temporary solution.
